icaruscode/PMT/CopyBeamTimePMTwaveforms_module.cc

```diff
@@ -476,20 +629,33 @@ auto icarus::CopyBeamTimePMTwaveforms::getReferenceTime
 
 
 //------------------------------------------------------------------------------
-bool icarus::CopyBeamTimePMTwaveforms::contains
-  (electronics_time time, raw::OpDetWaveform const& waveform) const
+bool icarus::CopyBeamTimePMTwaveforms::overlaps
+  (Interval_t const& time, raw::OpDetWaveform const& waveform) const
 {
   
   using util::quantities::points::microsecond;
   
   electronics_time const startWaveformTime
     { microsecond{ waveform.TimeStamp() } };
-  electronics_time const endWaveformTime
-    = startWaveformTime + fOpticalTick * waveform.size();
+    
+  Interval_t overlap
+    { startWaveformTime, startWaveformTime + fOpticalTick * waveform.size() };
+  overlap.intersect(time);
```

Comment:
What units are being used in this operation `startWaveformTime + fOpticalTick * waveform.size()`?
`startWaveformTime` is defined in microsecond a few lines above, but what about `fOpticalTick`? `detTimings.OpticalClockPeriod()` returns `auto` and `fOpticalTick` is defined as a `nanoseconds const`, so is it in nanoseconds? Am I missing a conversion somewhere?

Same question could be applied to the old code...

Reply:
Thank you for asking :sweat_smile:

`startWaveformTime` is a variable of `electronics_time` type, which is a time point derived from `util::quantities::points::microsecond`.
`fOpticalHits` is a time interval derived from `util::quantities::intervals::nanoseconds`.
These two are derivation of a [quantity type](https://nusoft.fnal.gov/larsoft/doxsvn/html/structutil_1_1quantities_1_1concepts_1_1Quantity.html#details) `second`; each of the types derived from [`second`](https://nusoft.fnal.gov/larsoft/doxsvn/html/namespaceutil_1_1quantities.html#ae3f21a6b9f866534e769a19192532eb4) are stored in a unit and can convert to any other. So if you have `microsecond{4} + nanosecond{20}` (or, with literals, `4_us + 20_ns`), it gets converted into a `microsecond` of value `4.02` (more precisely, the `operator+(microsecond, nanoseconds)` converts the second operand into `microseconds` and then does numerical addition.

You have been saved by a failed C&P on a more detailed description of quantities, intervals, points, and `detinfo::DetectorClocksWithUnits` and [`detinfo::DetectorTimings`](https://nusoft.fnal.gov/larsoft/doxsvn/html/classdetinfo_1_1DetectorTimings.html#details).

Reply:
It's clear now, thank you!